python/xir/ir.lark

```diff
@@ -4,11 +4,14 @@ include: "use" name ";"
 
 // the main circuit
 circuit: statement+
-statement: gatestmt | gate_def | operator_def | script_options
+statement: gatestmt | adjoint_gatestmt | ctrl_gatestmt | gate_def | operator_def | script_options
 gatestmt: name (params | options_dict) "|" wires ";"
 opstmt: expr "," obs_group";"
 script_options: _OPTIONS ":" (option ";")+ _END ";"
 
+adjoint_gatestmt: _ADJOINT gatestmt
+ctrl_gatestmt: _CTRL wires gatestmt
+
 gate_def: _GATE name params wires? ":" gatestmt+ _END ";"
```

Comment:
Should it be possible to include adjoint and control gates _within_ custom gates? E.g.
```
gate h(a)[0, 1]:
    adjoint rz(-2.3932854391951004) | [0];
    ctrl[0] rz(a) | [1];
end;
```
This one seems like a nice-to-have if it is not too complicated to add.

Reply:
This is a really good point. I think I might update the grammar on this to allow both that as well as any number of `ctrl` and `adjoint` modifiers on statements.

Reply:
Another datapoint: as a user, I would expect the grammar inside of a custom gate definition to exactly match what happens 'top level' when you are specifying gates to be executed!

Reply:
This now works as well! 🥳